README.md

This is an archive of libotr 3.2.1. I put it on Github because the official site
is down sometimes. This version is needed to run hexchat-otr.

Official source package is here: https://otr.cypherpunks.ca/libotr-3.2.1.tar.gz



ChangeLog
2012-08-14

	* src/auth.c: More carefully check for crazy short OTR auth
	messages.  Thanks to Ben Hawkes <hawkes@inertiawar.com> for the
	report.

2012-07-27

	* src/version.h: Update libotr version number to 3.2.1

2012-07-19

	* src/b64.[ch], src/proto.c, toolkit/parse.c: Clean up the
	previous b64 patch and apply it to all places where
	otrl_base64_decode() is called.

2012-07-17

	* src/b64.c: Use ceil instead of floor to compute the size
	of the data buffer.  This prevents a one-byte heap buffer
	overflow.  Thanks to Justin Ferguson <jnferguson@gmail.com>
	for the report.

2008-06-15:

	* README: Release version 3.2.0.

2008-06-13:

	* UPGRADING: Clarify what was new in 3.1.0, what was changed
	in 3.2.0.

2008-05-27:

	* UPGRADING: Update documentation.

	* README:
	* toolkit/*.[ch]:
	* src/*.[ch]: Update copyright dates to 2004-2008.

	* src/tlv.h: Add new OTRL_TLV_SMP1Q TLV type to indicate an
	instance of the first SMP message, with an explicit question.

	* src/sm.h:
	* src/sm.c: More carefully track the progress of the SMP using a
	new smp_prog_state field.  Also keep track of whether Bob
	received an explicit question from Alice using a new
	received_question field.

	* src/message.c: Handle explicit questions for the SMP.

	* src/message.c: Behave better if an SMP message fails
	verification.

	* README:
	* configure.ac:
	* src/version.h: Update version number to 3.2.0.

2007-07-26

	* src/sm.c:
	* src/message.c: ISO C cleanups (no mixing declarations with
	code)

	* src/sm.c: Fixed a 64-bit pointer error

2007-07-25

	* src/message.c: Behave sanely if we receive a totally malformed
	SMP message.

2007-07-24

	* src/proto.h:
	* src/proto.c:
	* src/message.c: Implemented fragmentation of large messages

	* src/message.h: New callback for fragmentation

	* src/privkey.h:
	* src/privkey.c (otrl_privkey_fingerprint_raw): New function to
	return a raw hash of an account's public key

	* src/proto.c: Keep track of the API version number passed to
	otrl_init()

	* src/context.h:
	* src/context.c:
	* src/tlv.h:
	* src/sm.h:
	* src/sm.c: Implemented the Socialist Millionaires' Protocol for
	authenticating buddies without using user-visible fingerprints

	* src/b64.h:
	* src/b64.c (decode, otrl_base64_decode): Corrected char vs.
	unsigned char

	* README:
	* configure.ac:
	* src/version.h: Change version number to 3.1.0

	* Most files: Update copyright information

2007-07-23

	* src/message.h:
	* src/message.c: Added account_name and account_name_free callbacks
	to OtrlMessageAppOps to let the application choose how to
	display the account name in OTR Error Messages.  Based on a
	patch from Evan Schoenberg <evan.s@dreskin.net>.

2006-07-24

	* src/privkey.h:
	* src/privkey.c: Add routines to read and write privkey and
	fingerprint data to FILE*s, instead of to filenames.

2006-05-09

	* Protocol-v2.html: Fix a typo, and correct the documentation
	regarding when MAC keys are revealed.

2006-04-13

	* src/context.h: Change "struct fingerprint" to "struct
	s_fingerprint" to appease some C++ compilers.

2006-02-09

	* src/auth.c (otrl_auth_handle_v1_key_exchange): Fix
	uninitialized variable received_pub.

2005-12-30

	* src/message.c: Fix a typo, thanks to Anton Blanchard
	<anton@samba.org>.

2005-11-20

	* src/proto.h: Fix typo in policy #defines.

2005-11-02

	* README:
	* src/version.h: Release version 3.0.0

2005-10-30

	* Protocol-v2.html: Clarified the uniqueness conditions for the
	counter.

	* src/auth.c (otrl_auth_handle_v1_key_exchange): Clear the auth
	structure when we receive an unexpected v1 Key Exchange Message.

2005-10-27

	* src/auth.h:
	* src/auth.c:
	* src/message.c: Ensure version 2 AKEs are always done with
	fresh D-H parameters.

	* src/proto.h:
	* src/proto.c:
	* src/message.c: Add a "flags" field to the version 2 Data
	Message, which can indicate that the Data Message should be
	ignored if unreadable (as opposed to displaying an error).

	* toolkit/parse.h:
	* toolkit/parse.c:
	* toolkit/otr_parse.c:
	* toolkit/otr_remac.c: Deal with the new kind of Data Message.

	* src/message.c: Use the gone_secure callback instead of the
	still_secure callback if the other side changes its fingerprint.

2005-10-19

	* src/context.h:
	* src/context.c: Added protocol_version as an explicit field in
	the ConnContext.

	* src/message.h:
	* src/message.c: protocol_version no longer needs to be
	explicitly passed to the gone_secure() and still_secure()
	callbacks.

	* packaging/fedora/libotr.spec: Patches from Paul

	* src/proto.c (rotate_dh_keys): Avoid potential double
	gcry_cipher_close().

	* src/tests.c: Regression test for double gcry_cipher_close().

2005-10-16

	* Major overhaul with implementation of version 2 AKE.

2005-08-08

	* toolkit/otr_parse.c (parse): Ignore MACs that are too short,
	rather than going into an infinite loop.

2005-08-04

	* Protocol: Added section describing fragments.

	* src/proto.h:
	* src/proto.c (otrl_proto_fragment_accumulate): 
	* src/context.h:
	* src/context.c (new_context, otrl_context_force_setup): Keep
	track of fragments in the ConnContext structure.

	* src/message.c (otrl_message_receiving): Handle fragments in
	received messages.

	* src/mem.c: Don't do arithmetic on void pointers.

2005-07-29

	* src/message.h:
	* src/message.c: Move ops to be the first param of
	new_fingerprint, as it is with all the other callbacks.

	* src/context.h:
	* src/context.c (otrl_context_set_preshared_secret):
	* src/dh.h:
	* src/dh.c (otrl_dh_session, otrl_dh_cmpctr):
	* src/message.h:
	* src/message.c (otrl_message_sending, send_or_error, process_kem)
	(otrl_message_receiving, otrl_message_disconnect):
	* src/privkey.h:
	* src/privkey.c (otrl_privkey_hash_to_human):
	* src/proto.h:
	* src/proto.c (otrl_proto_create_data):
	* src/tlv.h:
	* src/tlv.c (otrl_tlv_new, otrl_tlv_parse, otrl_tlv_seriallen)
	(otrl_tlv_serialize): Add missing "const"s.  (Closes #1243963)

2005-06-24

	* README:
	* configure.ac:
	* packaging/fedora/libotr.spec:

src/b64.c
/*
 *  Off-the-Record Messaging library
 *  Copyright (C) 2004-2008  Ian Goldberg, Chris Alexander, Nikita Borisov
 *                           <otr@cypherpunks.ca>
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of version 2.1 of the GNU Lesser General
 *  Public License as published by the Free Software Foundation.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* Modified from: */

/*********************************************************************\

MODULE NAME:    b64.c

AUTHOR:         Bob Trower 08/04/01

LICENCE:        Copyright (c) 2001 Bob Trower, Trantor Standard Systems Inc.

                Permission is hereby granted, free of charge, to any person
                obtaining a copy of this software and associated
                documentation files (the "Software"), to deal in the
                Software without restriction, including without limitation
                the rights to use, copy, modify, merge, publish, distribute,
                sublicense, and/or sell copies of the Software, and to
                permit persons to whom the Software is furnished to do so,
                subject to the following conditions:

                The above copyright notice and this permission notice shall
                be included in all copies or substantial portions of the
                Software.

                THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
                KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
                WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
                PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
                OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
                OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
                OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
                SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

VERSION HISTORY:
                Bob Trower 08/04/01 -- Create Version 0.00.00B

\******************************************************************* */

/* system headers */
#include <stdio.h>
#include <string.h>

/* libotr headers */
#include "b64.h"

/*
** Translation Table as described in RFC1113
*/
static const char cb64[]="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

/*
** Translation Table to decode (created by author)
*/
static const char cd64[]="|$$$}rstuvwxyz{$$$$$$$>?@ABCDEFGHIJKLMNOPQRSTUVW$$$$$$XYZ[\\]^_`abcdefghijklmnopq";

/*
** encodeblock
**
** encode up to 3 8-bit binary bytes as 4 '6-bit' characters.
** len must be 1, 2, or 3.
*/
static void encodeblock( char *out, const unsigned char *in, size_t len )
{
    unsigned char in0, in1, in2;
    in0 = in[0];
    in1 = len > 1 ? in[1] : 0;
    in2 = len > 2 ? in[2] : 0;

    out[0] = cb64[ in0 >> 2 ];
    out[1] = cb64[ ((in0 & 0x03) << 4) | ((in1 & 0xf0) >> 4) ];
    out[2] = len > 1 ? cb64[ ((in1 & 0x0f) << 2) | ((in2 & 0xc0) >> 6) ]
	             : '=';
    out[3] = len > 2 ? cb64[ in2 & 0x3f ]
	             : '=';
}

/*
 * base64 encode data.  Insert no linebreaks or whitespace.
 *
 * The buffer base64data must contain at least ((datalen+2)/3)*4 bytes of
 * space.  This function will return the number of bytes actually used.
 */
size_t otrl_base64_encode(char *base64data, const unsigned char *data,
	size_t datalen)
{
    size_t base64len = 0;

    while(datalen > 2) {
	encodeblock(base64data, data, 3);
	base64data += 4;
	base64len += 4;
	data += 3;
	datalen -= 3;
    }
    if (datalen > 0) {
	encodeblock(base64data, data, datalen);
	base64len += 4;
    }

    return base64len;
}

static size_t decode(unsigned char *out, const char *in, size_t b64len)
{   
    size_t written = 0;
    unsigned char c = 0;

    if (b64len > 0) {
	c = in[0] << 2;
    }
    if (b64len > 1) {
	out[0] = c | in[1] >> 4;
	written = 1;
	c = in[1] << 4;
    }
    if (b64len > 2) {
	out[1] = c | in[2] >> 2;
	written = 2;
	c = in[2] << 6;
    }
    if (b64len > 3) {
	out[2] = c | in[3];
	written = 3;
    }
    return written;
}

/*
 * base64 decode data.  Skip non-base64 chars, and terminate at the
 * first '=', or the end of the buffer.
 *
 * The buffer data must contain at least ((base64len+3) / 4) * 3 bytes
 * of space.  This function will return the number of bytes actually
 * used.
 */
size_t otrl_base64_decode(unsigned char *data, const char *base64data,
	size_t base64len)
{
    size_t datalen = 0;
    char b64[4];
    size_t b64accum = 0;

    while(base64len > 0) {
	char b = *base64data;
	unsigned char bdecode;
	++base64data;
	--base64len;
	if (b < '+' || b > 'z') continue;  /* Skip non-base64 chars */
	if (b == '=') {
	    /* Force termination */
	    datalen += decode(data, b64, b64accum);
	    base64len = 0;
	} else {
	    bdecode = cd64[b-'+'];
	    if (bdecode == '$') continue;  /* Skip non-base64 chars */
	    b64[b64accum++] = bdecode-'>';
	    if (b64accum == 4) {
		/* We have a complete block; decode it. */
		size_t written = decode(data, b64, b64accum);
		data += written;
		datalen += written;
		b64accum = 0;
	    }
	}
    }

    /* Just discard any short block at the end. */

    return datalen;
}

/*
 * Base64-encode a block of data, stick "?OTR:" and "." around it, and
 * return the result, or NULL in the event of a memory error.  The
 * caller must free() the return value.
 */
char *otrl_base64_otr_encode(const unsigned char *buf, size_t buflen)
{
    char *base64buf;
    size_t base64len;

    /* Make the base64-encoding. */
    base64len = ((buflen + 2) / 3) * 4;
    base64buf = malloc(5 + base64len + 1 + 1);
    if (base64buf == NULL) {
	return NULL;
    }
    memmove(base64buf, "?OTR:", 5);
    otrl_base64_encode(base64buf+5, buf, buflen);
    base64buf[5 + base64len] = '.';
    base64buf[5 + base64len + 1] = '\0';

    return base64buf;
}

/*
 * Base64-decode the portion of the given message between "?OTR:" and
 * ".".  Set *bufp to the decoded data, and set *lenp to its length.
 * The caller must free() the result.  Return 0 on success, -1 on a
 * memory error, or -2 on invalid input.
 */
int otrl_base64_otr_decode(const char *msg, unsigned char **bufp,
	size_t *lenp)
{
    char *otrtag, *endtag;
    size_t msglen, rawlen;
    unsigned char *rawmsg;

    otrtag = strstr(msg, "?OTR:");
    if (!otrtag) {
	return -2;
    }
    endtag = strchr(otrtag, '.');
    if (endtag) {
        msglen = endtag-otrtag;
    } else {
	return -2;
    }

    /* Skip over the "?OTR:" */
    otrtag += 5;
    msglen -= 5;

    /* Base64-decode the message */
    rawlen = OTRL_B64_MAX_DECODED_SIZE(msglen);   /* maximum possible */
    rawmsg = malloc(rawlen);
    if (!rawmsg && rawlen > 0) {
	return -1;
    }

    rawlen = otrl_base64_decode(rawmsg, otrtag, msglen);  /* actual size */

    *bufp = rawmsg;
    *lenp = rawlen;

    return 0;
}


toolkit/parse.c
/*
 *  Off-the-Record Messaging Toolkit
 *  Copyright (C) 2004-2008  Ian Goldberg, Chris Alexander, Nikita Borisov
 *                           <otr@cypherpunks.ca>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of version 2 of the GNU General Public License as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* system headers */
#include <stdio.h>
#include <stdlib.h>

/* libotr headers */
#include "b64.h"

/* toolkit headers */
#include "sha1hmac.h"
#include "parse.h"

/* Dump an unsigned int to a FILE * */
void dump_int(FILE *stream, const char *title, unsigned int val)
{
    fprintf(stream, "%s: %u\n", title, val);
}

/* Dump an mpi to a FILE * */
void dump_mpi(FILE *stream, const char *title, gcry_mpi_t val)
{
    size_t plen;
    unsigned char *d;
    
    gcry_mpi_print(GCRYMPI_FMT_USG, NULL, 0, &plen, val);
    d = malloc(plen);
    gcry_mpi_print(GCRYMPI_FMT_USG, d, plen, NULL, val);
    dump_data(stream, title, d, plen);
    free(d);
}

/* Dump data to a FILE * */
void dump_data(FILE *stream, const char *title, const unsigned char *data,
	size_t datalen)
{
    size_t i;
    fprintf(stream, "%s: ", title);
    for(i=0;i<datalen;++i) {
	fprintf(stream, "%02x", data[i]);
    }
    fprintf(stream, "\n");
}

/* base64 decode the message, and put the resulting size into *lenp */
static unsigned char *decode(const char *msg, size_t *lenp)
{
    const char *header, *footer;
    unsigned char *raw;
    size_t rawlen;

    /* Find the header */
    header = strstr(msg, "?OTR:");
    if (!header) return NULL;
    /* Skip the header */
    header += 5;

    /* Find the trailing '.' */
    footer = strchr(header, '.');
    if (!footer) footer = header + strlen(header);

    rawlen = OTRL_B64_MAX_DECODED_SIZE(footer-header);

    raw = malloc(rawlen);
    if (raw == NULL && rawlen > 0) return NULL;
    *lenp = otrl_base64_decode(raw, header, footer-header);

    return raw;
}

#define require_len(l) do { if (lenp < (l)) goto inv; } while(0)
#define read_int(x) do { \
	require_len(4); \
	(x) = (bufp[0] << 24) | (bufp[1] << 16) | (bufp[2] << 8 ) | bufp[3]; \
	bufp += 4; lenp -= 4; \
    } while(0)
#define read_mpi(x) do { \
	size_t mpilen; \
	read_int(mpilen); \
	require_len(mpilen); \
	gcry_mpi_scan(&(x), GCRYMPI_FMT_USG, bufp, mpilen, NULL); \
	bufp += mpilen; lenp -= mpilen; \
    } while(0)
#define read_raw(b, l) do { \
	require_len(l); \
	memmove((b), bufp, (l)); \
	bufp += (l); lenp -= (l); \
    } while(0)
#define write_int(x) do { \
	bufp[0] = ((x) >> 24) & 0xff; \
	bufp[1] = ((x) >> 16) & 0xff; \
	bufp[2] = ((x) >> 8) & 0xff; \
	bufp[3] = (x) & 0xff; \
	bufp += 4; lenp -= 4; \
    } while(0)
#define write_mpi(x,l) do { \
	write_int(l); \
	gcry_mpi_print(GCRYMPI_FMT_USG, bufp, lenp, NULL, (x)); \
	bufp += (l); lenp -= (l); \
    } while(0)
#define write_raw(x,l) do { \
	memmove(bufp, (x), (l)); \
	bufp += (l); lenp -= (l); \
    } while(0)

/* Parse a Key Exchange Message into a newly-allocated KeyExchMsg structure */
KeyExchMsg parse_keyexch(const char *msg)
{
    KeyExchMsg kem = NULL;
    size_t lenp;
    unsigned char *raw = decode(msg, &lenp);
    unsigned char *bufp = raw;
    if (!raw) goto inv;

    kem = calloc(1, sizeof(struct s_KeyExchMsg));
    if (!kem) {
	free(raw);
	goto inv;
    }

    kem->raw = raw;
    kem->sigstart = bufp;

    require_len(3);
    if (memcmp(bufp, "\x00\x01\x0a", 3)) goto inv;
    bufp += 3; lenp -= 3;

    require_len(1);
    kem->reply = *bufp;
    bufp += 1; lenp -= 1;

    read_mpi(kem->p);
    read_mpi(kem->q);
    read_mpi(kem->g);
    read_mpi(kem->e);

    read_int(kem->keyid);

    read_mpi(kem->y);

    kem->sigend = bufp;

    require_len(40);
    gcry_mpi_scan(&kem->r, GCRYMPI_FMT_USG, bufp, 20, NULL);
    gcry_mpi_scan(&kem->s, GCRYMPI_FMT_USG, bufp+20, 20, NULL);
    bufp += 40; lenp -= 40;

    if (lenp != 0) goto inv;

    return kem;
inv:
    free_keyexch(kem);
    return NULL;
}

/* Deallocate a KeyExchMsg and all of the data it points to */
void free_keyexch(KeyExchMsg keyexch)
{
    if (!keyexch) return;
    free(keyexch->raw);
    gcry_mpi_release(keyexch->p);
    gcry_mpi_release(keyexch->q);
    gcry_mpi_release(keyexch->g);
    gcry_mpi_release(keyexch->e);
    gcry_mpi_release(keyexch->y);
    gcry_mpi_release(keyexch->r);
    gcry_mpi_release(keyexch->s);
    free(keyexch);
}

/* Parse a D-H Commit Message into a newly-allocated CommitMsg structure */
CommitMsg parse_commit(const char *msg)
{
    CommitMsg cmsg = NULL;
    size_t lenp;
    unsigned char *raw = decode(msg, &lenp);
    unsigned char *bufp = raw;
    if (!raw) goto inv;

    cmsg = calloc(1, sizeof(struct s_CommitMsg));
    if (!cmsg) {
	free(raw);
	goto inv;
    }

    cmsg->raw = raw;

    require_len(3);
    if (memcmp(bufp, "\x00\x02\x02", 3)) goto inv;
    bufp += 3; lenp -= 3;

    read_int(cmsg->enckeylen);
    cmsg->enckey = malloc(cmsg->enckeylen);
    if (!cmsg->enckey && cmsg->enckeylen > 0) goto inv;
    read_raw(cmsg->enckey, cmsg->enckeylen);

    read_int(cmsg->hashkeylen);
    cmsg->hashkey = malloc(cmsg->hashkeylen);
    if (!cmsg->hashkey && cmsg->hashkeylen > 0) goto inv;
    read_raw(cmsg->hashkey, cmsg->hashkeylen);

    if (lenp != 0) goto inv;

    return cmsg;
inv:
    free_commit(cmsg);
    return NULL;
}

/* Deallocate a CommitMsg and all of the data it points to */
void free_commit(CommitMsg cmsg)
{
    if (!cmsg) return;
    free(cmsg->raw);
    free(cmsg->enckey);
    free(cmsg->hashkey);
    free(cmsg);
}

/* Parse a D-H Key Message into a newly-allocated KeyMsg structure */
KeyMsg parse_key(const char *msg)
{
    KeyMsg kmsg = NULL;
    size_t lenp;
    unsigned char *raw = decode(msg, &lenp);
    unsigned char *bufp = raw;
    if (!raw) goto inv;

    kmsg = calloc(1, sizeof(struct s_KeyMsg));
    if (!kmsg) {
	free(raw);
	goto inv;
    }

    kmsg->raw = raw;

    require_len(3);
    if (memcmp(bufp, "\x00\x02\x0a", 3)) goto inv;
    bufp += 3; lenp -= 3;

    read_mpi(kmsg->y);

    if (lenp != 0) goto inv;

    return kmsg;
inv:
    free_key(kmsg);
    return NULL;
}

/* Deallocate a KeyMsg and all of the data it points to */
void free_key(KeyMsg kmsg)
{
    if (!kmsg) return;
    free(kmsg->raw);
    gcry_mpi_release(kmsg->y);
    free(kmsg);
}

/* Parse a Reveal Signature Message into a newly-allocated RevealSigMsg
 * structure */
RevealSigMsg parse_revealsig(const char *msg)
{
    RevealSigMsg rmsg = NULL;
    size_t lenp;
    unsigned char *raw = decode(msg, &lenp);
    unsigned char *bufp = raw;
    if (!raw) goto inv;

    rmsg = calloc(1, sizeof(struct s_RevealSigMsg));
    if (!rmsg) {
	free(raw);
	goto inv;
    }

    rmsg->raw = raw;

    require_len(3);
    if (memcmp(bufp, "\x00\x02\x11", 3)) goto inv;
    bufp += 3; lenp -= 3;

    read_int(rmsg->keylen);
    rmsg->key = malloc(rmsg->keylen);
    if (!rmsg->key && rmsg->keylen > 0) goto inv;
    read_raw(rmsg->key, rmsg->keylen);

    read_int(rmsg->encsiglen);
    rmsg->encsig = malloc(rmsg->encsiglen);
    if (!rmsg->encsig && rmsg->encsiglen > 0) goto inv;
    read_raw(rmsg->encsig, rmsg->encsiglen);

    read_raw(rmsg->mac, 20);

    if (lenp != 0) goto inv;

    return rmsg;
inv:
    free_revealsig(rmsg);
    return NULL;
}

/* Deallocate a RevealSigMsg and all of the data it points to */
void free_revealsig(RevealSigMsg rmsg)
{
    if (!rmsg) return;
    free(rmsg->raw);
    free(rmsg->key);
    free(rmsg->encsig);
    free(rmsg);
}

/* Parse a Signature Message into a newly-allocated SignatureMsg structure */
SignatureMsg parse_signature(const char *msg)
{
    SignatureMsg smsg = NULL;
    size_t lenp;
    unsigned char *raw = decode(msg, &lenp);
    unsigned char *bufp = raw;
    if (!raw) goto inv;

    smsg = calloc(1, sizeof(struct s_SignatureMsg));
    if (!smsg) {
	free(raw);
	goto inv;
    }

    smsg->raw = raw;

    require_len(3);
    if (memcmp(bufp, "\x00\x02\x12", 3)) goto inv;
    bufp += 3; lenp -= 3;

    read_int(smsg->encsiglen);
    smsg->encsig = malloc(smsg->encsiglen);
    if (!smsg->encsig && smsg->encsiglen > 0) goto inv;
    read_raw(smsg->encsig, smsg->encsiglen);

    read_raw(smsg->mac, 20);

    if (lenp != 0) goto inv;

    return smsg;
inv:
    free_signature(smsg);
    return NULL;
}

/* Deallocate a SignatureMsg and all of the data it points to */
void free_signature(SignatureMsg smsg)
{
    if (!smsg) return;
    free(smsg->raw);
    free(smsg->encsig);
    free(smsg);
}

/* Parse a Data Message into a newly-allocated DataMsg structure */
DataMsg parse_datamsg(const char *msg)
{
    DataMsg datam = NULL;
    size_t lenp;
    unsigned char *raw = decode(msg, &lenp);
    unsigned char *bufp = raw;
    unsigned char version;
    if (!raw) goto inv;

    datam = calloc(1, sizeof(struct s_DataMsg));
    if (!datam) {
	free(raw);
	goto inv;
    }

    datam->raw = raw;
    datam->rawlen = lenp;
    datam->macstart = bufp;

    require_len(3);
    if (memcmp(bufp, "\x00\x01\x03", 3) && memcmp(bufp, "\x00\x02\x03", 3))
	goto inv;
    version = bufp[1];
    bufp += 3; lenp -= 3;

    if (version == 2) {
	require_len(1);
	datam->flags = bufp[0];
	bufp += 1; lenp -= 1;
    } else {
	datam->flags = -1;
    }
    read_int(datam->sender_keyid);
    read_int(datam->rcpt_keyid);
    read_mpi(datam->y);
    read_raw(datam->ctr, 8);
    read_int(datam->encmsglen);
    datam->encmsg = malloc(datam->encmsglen);
    if (!datam->encmsg && datam->encmsglen > 0) goto inv;
    read_raw(datam->encmsg, datam->encmsglen);
    datam->macend = bufp;
    read_raw(datam->mac, 20);
    read_int(datam->mackeyslen);
    datam->mackeys = malloc(datam->mackeyslen);
    if (!datam->mackeys && datam->mackeyslen > 0) goto inv;
    read_raw(datam->mackeys, datam->mackeyslen);

    if (lenp != 0) goto inv;

    return datam;
inv:
    free_datamsg(datam);
    return NULL;
}

/* Recalculate the MAC on the message, base64-encode the resulting MAC'd
 * message, and put on the appropriate header and footer.  Return a
 * newly-allocated pointer to the result, which the caller will have to
 * free(). */
char *remac_datamsg(DataMsg datamsg, unsigned char mackey[20])
{
    size_t rawlen, lenp;
    size_t ylen;
    size_t base64len;
    char *outmsg;
    unsigned char *raw, *bufp;
    unsigned char version = (datamsg->flags >= 0 ? 2 : 1);
    
    /* Calculate the size of the message that will result */
    gcry_mpi_print(GCRYMPI_FMT_USG, NULL, 0, &ylen, datamsg->y);
    rawlen = 3 + (version == 2 ? 1 : 0) + 4 + 4 + 4 + ylen + 8 + 4 +
	datamsg->encmsglen + 20 + 4 + datamsg->mackeyslen;

    /* Construct the new raw message (note that some of the pieces may
     * have been altered, so we construct it from scratch). */
    raw = malloc(rawlen);
    if (!raw) {
	fprintf(stderr, "Out of memory!\n");
	exit(1);
    }
    bufp = raw;
    lenp = rawlen;
    datamsg->macstart = raw;
    datamsg->macend = NULL;
    free(datamsg->raw);
    datamsg->raw = raw;
    datamsg->rawlen = rawlen;

    if (version == 1) {
	memmove(bufp, "\x00\x01\x03", 3);
    } else {
	memmove(bufp, "\x00\x02\x03", 3);
    }
    bufp += 3; lenp -= 3;
    if (version == 2) {
	bufp[0] = datamsg->flags;
	bufp += 1; lenp -= 1;
    }
    write_int(datamsg->sender_keyid);
    write_int(datamsg->rcpt_keyid);
    write_mpi(datamsg->y, ylen);
    write_raw(datamsg->ctr, 8);
    write_int(datamsg->encmsglen);
    write_raw(datamsg->encmsg, datamsg->encmsglen);
    datamsg->macend = bufp;

    /* Recalculate the MAC */
    sha1hmac(datamsg->mac, mackey, datamsg->macstart,
	    datamsg->macend - datamsg->macstart);

    write_raw(datamsg->mac, 20);
    write_int(datamsg->mackeyslen);
    write_raw(datamsg->mackeys, datamsg->mackeyslen);
    
    if (lenp != 0) {
	fprintf(stderr, "Error creating OTR Data Message.\n");
	exit(1);
    }

    base64len = 5 + ((datamsg->rawlen + 2) / 3) * 4 + 1 + 1;
    outmsg = malloc(base64len);
    if (!outmsg) return NULL;

    memmove(outmsg, "?OTR:", 5);
    otrl_base64_encode(outmsg + 5, datamsg->raw, datamsg->rawlen);
    strcpy(outmsg + base64len - 2, ".");
    return outmsg;
}

/* Assemble a new Data Message from its pieces.  Return a
 * newly-allocated string containing the base64 representation. */
char *assemble_datamsg(unsigned char mackey[20], int flags,
	unsigned int sender_keyid, unsigned int rcpt_keyid, gcry_mpi_t y,
	unsigned char ctr[8], unsigned char *encmsg, size_t encmsglen,
	unsigned char *mackeys, size_t mackeyslen)
{
    DataMsg datam = calloc(1, sizeof(struct s_DataMsg));
    char *newmsg = NULL;
    if (!datam) goto inv;
    datam->flags = flags;
    datam->sender_keyid = sender_keyid;
    datam->rcpt_keyid = rcpt_keyid;
    datam->y = gcry_mpi_copy(y);
    memmove(datam->ctr, ctr, 8);
    datam->encmsg = malloc(encmsglen);
    if (!datam->encmsg && encmsglen > 0) goto inv;
    memmove(datam->encmsg, encmsg, encmsglen);
    datam->encmsglen = encmsglen;
    datam->mackeys = malloc(mackeyslen);
    if (!datam->mackeys && mackeyslen > 0) goto inv;
    memmove(datam->mackeys, mackeys, mackeyslen);
    datam->mackeyslen = mackeyslen;

    /* Recalculate the MAC and base64-encode the result */
    newmsg = remac_datamsg(datam, mackey);
    free_datamsg(datam);
    return newmsg;
inv:
    free_datamsg(datam);
    return NULL;
}

/* Deallocate a DataMsg and all of the data it points to */
void free_datamsg(DataMsg datamsg)
{
    if (!datamsg) return;
    free(datamsg->raw);
    gcry_mpi_release(datamsg->y);
    free(datamsg->encmsg);
    free(datamsg->mackeys);
    free(datamsg);
}

static int ctoh(char c)
{
    if (c >= '0' && c <= '9') return (c-'0');
    if (c >= 'a' && c <= 'f') return (c-'a'+10);
    if (c >= 'A' && c <= 'F') return (c-'A'+10);
    return -1;
}

/* Convert a string of hex chars to a buffer of unsigned chars. */
void argv_to_buf(unsigned char **bufp, size_t *lenp, char *arg)
{
    unsigned char *buf;
    size_t len, i;

    *bufp = NULL;
    *lenp = 0;

    len = strlen(arg);
    if (len % 2) {
	fprintf(stderr, "Argument ``%s'' must have even length.\n", arg);
	return;
    }
    buf = malloc(len/2);
    if (buf == NULL && len > 0) {
	fprintf(stderr, "Out of memory!\n");
	return;
    }

    for(i=0;i<len/2;++i) {
	int hi = ctoh(arg[2*i]);
	int lo = ctoh(arg[2*i+1]);
	if (hi < 0 || lo < 0) {
	    free(buf);
	    fprintf(stderr, "Illegal hex char in argument ``%s''.\n", arg);
	    return;
	}
	buf[i] = (hi << 4) + lo;
    }
    *bufp = buf;
    *lenp = len/2;
}
